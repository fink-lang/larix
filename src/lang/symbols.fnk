{symbol} = import '@fink/prattler/symbols'
{curr_loc} = import '@fink/prattler'

{next_is_new_expr} = import './block/indentation'
{single_expression} = import './block/expr'



infix_led = fn type, op, lbp: fn ctx, left:
  {loc: {start}} = left
  [right, next_ctx] = single_expression:: ctx, lbp
  {loc: {end}} = right

  [{type, op, left, right, loc: {start, end}}, next_ctx]


left_binding = fn lbp: fn ctx:
  match ctx:
    next_is_new_expr:: ?:
      0
    else:
      lbp


non_bining = fn: fn: 0


infix = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp + 1
  }


infix_right = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp - 1
  }


# TODO: why is this not type, op?
prefix = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn lbp: fn ctx:
      {start} = curr_loc:: ctx
      [right, next_ctx] = single_expression:: ctx, lbp
      {end} = right.loc

      [{type, op, right, loc: {start, end}}, next_ctx]
  }


literal = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn: fn ctx:
      loc = curr_loc:: ctx
      # value = curr_value:: ctx
      [{type, value: op, loc}, ctx]
  }


terminator = fn op:
  {
    ...symbol:: op

    lbp: left_binding
  }
