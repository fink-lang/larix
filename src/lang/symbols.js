import {symbol as parser_symbol} from '@fink/prattler/symbols';
import {next_loc} from '@fink/prattler';
import {indentation} from './indentation';


export const symbol = (id)=> ({
  ...parser_symbol(id),

  lbp: (lbp)=> (ctx)=> {
    if (next_loc(ctx).start.column <= indentation(ctx)) {
      return 0;
    }
    return lbp;
  }
});
