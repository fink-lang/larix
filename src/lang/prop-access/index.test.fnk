{describe, it, expect, to_match_snapshot} = import '@fink/jest'

{parse, parse_expr} = import '../../'


describe 'prop access', fn:

  it 'parses single line', fn:
    expect
      parse_expr `foo.bar`
      to_match_snapshot


  it 'parses with taken keywords', fn:
    expect
      parse_expr `foo.match`
      to_match_snapshot

    expect
      parse_expr `foo.fn`
      to_match_snapshot


  it `parses member expr: foo.'bar spam'`, fn:
    expect
      parse_expr `foo.'bar spam'`
      to_match_snapshot


  it 'parses member expr: foo.(Symbol.iterator)', fn:
    expect
      parse_expr 'foo.(Symbol.iterator)'
      to_match_snapshot


  it 'errors with invalid member)', fn:
    expect
      parse 'foo.[bar]', 'test.fnk'
      to_match_snapshot

