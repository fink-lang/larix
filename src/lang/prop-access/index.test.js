import {other_token} from '@fink/prattler/symbols';

import {parse_expr} from '../../';


describe('prop access', ()=> {

  it('parses single line: foo.bar', ()=> {
    expect(
      parse_expr(`foo.bar`)
    ).toMatchSnapshot();
  });


  it('parses member expr: foo.`bar spam`', ()=> {
    expect(
      parse_expr('foo.`bar spam`')
    ).toMatchSnapshot();
  });


  it('parses member expr: foo.(Symbol.iterator)', ()=> {
    expect(
      parse_expr('foo.(Symbol.iterator)')
    ).toMatchSnapshot();
  });
});
