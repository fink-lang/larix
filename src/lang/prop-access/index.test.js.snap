// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prop access parses member expr: foo.(Symbol.iterator) 1`] = `
member . (1:0-1:21)
  other (1:0-1:3) foo
  group (1:4-1:21)
    member . (1:5-1:20)
      other (1:5-1:11) Symbol
      other (1:12-1:20) iterator
`;

exports[`prop access parses member expr: foo.\`bar spam\` 1`] = `
member . (1:0-1:14)
  other (1:0-1:3) foo
  string \` (1:4-1:14)
    \`bar spam\`
`;

exports[`prop access parses single line: foo.bar 1`] = `
member . (1:0-1:7)
  other (1:0-1:3) foo
  other (1:4-1:7) bar
`;
